### PR TITLE
Add support for command line args, update application shims.

### PR DESCRIPTION
**Pull request by gspencergoog**
This adds support for command line args on Linux and macOS, and updates the application shims to the latest version.

/cc @yjbanov

### CODE REVIEW

File: bin/generate.dart [hunk 1]
```diff
@@ -106,17 +100,20 @@ class DiagramGenerator {
...
-    final List<File> outputFiles = await _combineAnimations(await _transferImages());
+    final List<File> outputFiles =
+        await _combineAnimations(await _transferImages());
     await _optimizeImages(outputFiles);
     if (cleanup) {
       await temporaryDirectory.delete(recursive: true);
     }
-    print('Elapsed time for diagram generation: ${DateTime.now().difference(startTime)}');
+    print(
+        'Elapsed time for diagram generation: ${DateTime.now().difference(startTime)}');
```
Comment (Hixie) [inline bin/generate.dart:116]: dartfmt certainly valiantly saves us from going over 80 characters here

File: bin/generate.dart [hunk 2]
```diff
@@ -130,17 +127,52 @@ class DiagramGenerator {
...
+          ? <String>[
+              '--route',
+              'args:${Uri.encodeComponent(filters.join(' '))}'
+            ]
+          : <String>[];
+    } else {
+      filterArgs = <String>[];
+      for (final String arg in filters) {
+        filterArgs.add('--dart-entrypoint-args');
+        filterArgs.add(arg);
```
Comment (Hixie) [inline bin/generate.dart:165]: we really should just support --route everywhere... (but that's an issue for another PR)

File: bin/generate.dart [hunk 3]
```diff
@@ -264,13 +296,14 @@ class DiagramGenerator {
...
 
   Future<List<File>> _combineAnimations(List<File> inputFiles) async {
-    print('Processing ${inputFiles.length} files...');
     final List<File> errorFiles = inputFiles
       .where((File input) => path.basename(input.path) == 'error.log')
       .toList();
 
     if (errorFiles.length != 1)
-      throw GeneratorException('Subprocess did not complete cleanly!');
+      throw GeneratorException('Subprocess did not complete cleanly!: $errorFiles');
```
Comment (Hixie) [inline bin/generate.dart:304]: did you mean to include $errorFiles in the output here?

File: bin/generate.dart [hunk 4]
```diff
@@ -353,23 +386,87 @@ class DiagramGenerator {
...
+// Android device on Windows.
+String getDefaultDevice(Map<String, Map<String, String>> devices) {
+  if (devices.isEmpty) {
+    return '';
+  }
+  final String platform = Platform.operatingSystem.toLowerCase();
+  if (devices.keys.contains(platform)) {
+    if (platform != 'windows') {
+      // We don't yet support Windows as a device, since the process requires
+      // some Unix tools still (ffmpeg, optipng, etc.).
```
Comment (Hixie) [inline bin/generate.dart:428]: i don't understand the situation on windows. if it needs tools, shouldn't we just not do anything? what does it mean to default to anything on windows?

File: bin/generate.dart [hunk 5]
```diff
@@ -353,23 +386,87 @@ class DiagramGenerator {
...
+      abbr: 't',
+      help: 'Specify a temporary directory to use (implies --keep-tmp)');
   parser.addOption('device-id',
-      abbr: 'd', help: 'Specify a device to use for generating the diagrams', defaultsTo: 'linux');
+      abbr: 'd',
+      help: 'Specify a device ID to use for generating the diagrams. Defaults '
+          'to the host platform that the script is run on, if that platform is '
+          'supported, or an attached device if not. Default on this machine is '
+          '${getDefaultDevice(devices)}. Available devices are:\n${getDeviceList(devices)}\n',
+      defaultsTo: getDefaultDevice(devices));
```
Comment (Hixie) [inline bin/generate.dart:458]: no need to say what the default is in the help string, args prints it out regardless

File: utils/diagram_generator/web/index.html [hunk 6]
```diff
@@ -0,0 +1,98 @@
+<!DOCTYPE html>
```
Comment (Hixie) [inline utils/diagram_generator/web/index.html:1]: not sure there's much point having the web version :-)

**Review by Hixie — APPROVED** [review]
LGTM modulo some minor stuff

File: utils/diagram_generator/web/index.html (hunk 6 above)
Comment (gspencergoog) [reply utils/diagram_generator/web/index.html]: Yeah, good point.  I just had the option enabled when I generated the shims.  Removed this dir.

**Pull request merged**